packages/protocol/contracts/ACE/validators/privateRange/PrivateRange.sol

```diff
@@ -0,0 +1,355 @@
+                let b := add(0x2c0, mul(n, 0x80))
+
+                // Iterate over every note and calculate the blinding factor B_i = \gamma_i^{kBar}h^{aBar}\sigma_i^{-c}.
+                // We use the AZTEC protocol pairing optimization to reduce the number of pairing comparisons to 1,
+                //  which adds some minor alterations
+
+                // Iterate over every note and calculate the blinding factor B_i = \gamma_i^{kBar}h^{aBar}\sigma_i^{-c}.
+                // We use the AZTEC protocol pairing optimization to reduce the number of pairing comparisons to 1,
+                //  which adds some minor alterations
+                for { let i := 0 } lt(i, n) { i := add(i, 0x01) } {
+
+                    // Get the calldata index of this note
+                    let noteIndex := add(add(notes, 0x20), mul(i, 0xc0))
+
+                    // Define variables k, a and c.
+                    // If i <= m then
+                    //   k = kBar_i
+                    //   a = aBar_i
+                    //   c = challenge
+                    // If i > m then we add a modification for the pairing optimization
+                    //   k = kBar_i * x_i
+                    //   a = aBar_i * x_i
+                    //   c = challenge * x_i
+                    // Set j = i - (m + 1).
+                    // x_0 = 1
+                    // x_1 = keccak256(input string)
+                    // all other x_{j} = keccak256(x_{j-1})
+                    // The reason for doing this is that the point  \sigma_i^{-cx_j} can be re-used in the pairing check
+                    // Instead of validating e(\gamma_i, t_2) == e(\sigma_i, g_2) for all i = [m+1,\ldots,n]
+                    // We validate e(\Pi_{i=m+1}^{n}\gamma_i^{-cx_j}, t_2) == e(\Pi_{i=m+1}^{n}\sigma_i^{cx_j}, g_2).
+                    // x_j is a pseudorandom variable whose entropy source is the input string, allowing for
+                    // a sum of commitment points to be evaluated in one pairing comparison
+                    let k := calldataload(noteIndex)
+                    let a := calldataload(add(noteIndex, 0x20))
+                    let c := challenge
+
+                    switch gt(i, 1)
+                    case 1 {
+
+                        // k_3 = k_1 - k_2
+                        k := addmod(
+                                calldataload(sub(noteIndex, 0x180)), // k_1
+                                sub(
+                                    gen_order,
+                                    calldataload(sub(noteIndex, 0xc0))), // k_2x§
+                                gen_order)
+                    } 
+                    case 0 {
+                        k := calldataload(noteIndex)
+                    }
+
+                    // Check this commitment is well formed...
+                    validateCommitment(noteIndex, k, a)
+
+                    // Set k = kx_j, a = ax_j, c = cx_j, where j = i - (m+1)
+                    if gt(i, 0x00) {
```

Comment:
(small optimiation, but we can replace `if gt(i, 0x00)` with `if i`. The `jumpi(x, location)` opcode will execute the jump instruction if `x` is anything but zero.

Reply:
Nice, done